### PR TITLE
Draft review of dotnet guide for code coverage import destined for Community guides

### PR DESCRIPTION
**Pull request by joe-tingsanchali-sonarsource**
Draft review of dotnet guide for code coverage import destined for Community guides.

### CODE REVIEW

File: dotnet-coverage-test-data-import-guide.md [hunk 1]
```diff
@@ -24,15 +24,15 @@ Second, when importing code coverage reports, you need to ensure that you specif
...
+Third, understand the basics of...
+* SonarScanner for MSBuild by reading its docs [here](https://docs.sonarqube.org/latest/analysis/scan/sonarscanner-for-msbuild/), in case you are using it in your pipeline.
+* Azure Devops - understand the integration between Azure DevOps and SonarCloud by reading [this guideline](https://azuredevopslabs.com//labs/vstsextend/sonarcloud/). =====> will add the reference to the yaml file for “Prepare SonarCloudAnalysis” step......
 
 #### How to Get Debug Level Logs
 
-Add `/d:sonar.verbose=true` or `/d:"sonar.verbose=true"` to the
- SonarScanner.MSBuild.exe `begin` command to get more detailed logs. Usually, you will find your error here!
-`SonarCloudPrepare` task in case you are using Azure DevOps
+Add `/d:sonar.verbose=true` or `/d:"sonar.verbose=true"` to the...
```
Comment (okorach) [inline dotnet-coverage-test-data-import-guide.md]: Pick a single syntax here: I'd suggest to always use /d:"\<property\>=\<value\>" everywhere in the doc. That's the syntax that always work for any property

File: dotnet-coverage-test-data-import-guide.md [hunk 2]
```diff
@@ -24,15 +24,15 @@ Second, when importing code coverage reports, you need to ensure that you specif
 * sonar.vbnet.opencover.reportsPaths
     * Path to OpenCover coverage report.
 
-Third, understand the basics of 
-SonarScanner for MSBuild by reading its docs [here](https://docs.sonarqube.org/latest/analysis/scan/sonarscanner-for-msbuild/), in case you are using it in your pipeline.
-Azure Devops - understand the integration between Azure DevOps and SonarCloud by reading [this guideline](https://azuredevopslabs.com//labs/vstsextend/sonarcloud/)..
+Third, understand the basics of...
+* SonarScanner for MSBuild by reading its docs [here](https://docs.sonarqube.org/latest/analysis/scan/sonarscanner-for-msbuild/), in case you are using it in your pipeline.
```
Comment (okorach) [inline dotnet-coverage-test-data-import-guide.md]: I'd anticipate the imminent change and use the work SonarScanner for .NET (formerly SonarScanner for MSBuild)

File: dotnet-coverage-test-data-import-guide.md [hunk 3]
```diff
@@ -24,15 +24,15 @@ Second, when importing code coverage reports, you need to ensure that you specif
...
+* Azure Devops - understand the integration between Azure DevOps and SonarCloud by reading [this guideline](https://azuredevopslabs.com//labs/vstsextend/sonarcloud/). =====> will add the reference to the yaml file for “Prepare SonarCloudAnalysis” step......
```
Comment (okorach) [inline dotnet-coverage-test-data-import-guide.md]: Comment valid for several areas in the guide. Refer to the SonarCloud AND SonarQube equivalent reference (ie also mention SonarQubePrepareAnalysis on this case)

File: dotnet-coverage-test-data-import-guide.md [hunk 4]
```diff
@@ -24,15 +24,15 @@ Second, when importing code coverage reports, you need to ensure that you specif
...
 
-Add `/d:sonar.verbose=true` or `/d:"sonar.verbose=true"` to the
- SonarScanner.MSBuild.exe `begin` command to get more detailed logs. Usually, you will find your error here!
-`SonarCloudPrepare` task in case you are using Azure DevOps
+Add `/d:sonar.verbose=true` or `/d:"sonar.verbose=true"` to the...
+* SonarScanner.MSBuild.exe `begin` command to get more detailed logs ======>should mention the alternative for Azure Devops integration, where the Scanner for MSBuild is hidden
+* `SonarCloudPrepare` task in case you are using Azure DevOps
 
 For example:
 `SonarScanner.MSBuild.exe begin /k:“MyProject” /d:sonar.verbose=true`
```
Comment (okorach) [inline dotnet-coverage-test-data-import-guide.md]: /d:"sonar.verbose=true"

**Review by okorach — CHANGES_REQUESTED** [review]
Can't add comments on areas of the code you did not change so here it is:


*1. Add example of right usage with comma separated values*
:o: `/d:"sonar.cs.vscoveragexml.reportsPaths=coverage/joe*.coveragexml,otherdir/olivier*.coveragexml"

*2. Example of wrong usage*

Wrong :x:
```
# Don't point at directories but files
-d:"sonar.cs.vscoveragexml.reportsPaths=coverage/"

# Don't use absolute paths, favor paths relative to the project root directory
-d:"sonar.cs.vscoveragexml.reportsPaths=C:\users\joe\project\coverage\*.xml"

# Don't point files that are not under the project root directory, always generate coverage under this home directry
-d:"sonar.cs.vscoveragexml.reportsPaths=..\coverage\*.xml"
```

*3. Careful to cover SonarCloud and SonarQube case everywhere (it's too SonarCloud biased right now)*

*3. Careful to not be too MSBuild biased, cover also dotnet core (becoming more and more popular)*

*4. I feel the order of steps in not so clear in  Prematurely running report generation*
I'd write:
The report generation process must be executed after the **build** step and before the **Scanner end** step. The following steps detail importing .NET reports:

*5. Careful, if you also decide to cover VB.Net to include it everywhere, eg
```
DEBUG: '<PATH\FILE.cs>' indexed with language 'cs'
DEBUG: '<PATH\FILE.vb>' indexed with language 'vb'
```

Cheers

File: dotnet-coverage-test-data-import-guide.md (hunk 3 above)
Comment (joe-tingsanchali-sonarsource) [reply dotnet-coverage-test-data-import-guide.md]: Yea, this is my weak point: I haven't used SonarCloud enough to see the equivalence. I'll have to play with SonarCloud more.

File: dotnet-coverage-test-data-import-guide.md (hunk 2 above)
Comment (joe-tingsanchali-sonarsource) [reply dotnet-coverage-test-data-import-guide.md]: Good point. I'll reference both then.

File: dotnet-coverage-test-data-import-guide.md (hunk 1 above)
Comment (joe-tingsanchali-sonarsource) [reply dotnet-coverage-test-data-import-guide.md]: True. Perhaps I can make a single sentence that mentions the equivalent syntax but stick with the double-quotes throughout?

EDIT: nevermind, I'll just stick to one syntax throughout.

File: dotnet-coverage-test-data-import-guide.md [hunk 5]
```diff
@@ -1,117 +1,137 @@
...
+    * sonar.cs.vscoveragexml.reportsPaths
+        * Path to Visual Studio Code Coverage report. Multiple paths may be comma-delimited, or included via wildcards.
+        * Ensure that you have converted the code coverage report from Binary into XML (see "Convert the Code Coverage Report from Binary into XML" section on our [[Coverage & Test Data] Generate Reports for C#, VB.net](https://community.sonarsource.com/t/coverage-test-data-generate-reports-for-c-vb-net/9871) Community guide)
+    * sonar.cs.dotcover.reportsPaths
+        * Path to dotCover coverage report.
+    * sonar.cs.opencover.reportsPaths
+        * Path to OpenCover coverage report.
+    * sonar.vbnet.dotcover.reportsPaths
+        * Path to dotCover coverage report.
+    * sonar.vbnet.opencover.reportsPaths
```
Comment (andrei-epure-sonarsource) [inline dotnet-coverage-test-data-import-guide.md]: also `sonar.vbnet.vscoveragexml.reportsPaths`

File: dotnet-coverage-test-data-import-guide.md [hunk 6]
```diff
@@ -1,117 +1,137 @@
...
+    |---|---|
+    |?|a single character|
+    |*|any number of characters|
+    |**|any number of directories|
+    
+    Here is a resource to understand some other examples: [Pattern matching guide](https://confluence.atlassian.com/fisheye/pattern-matching-guide-960155410.html)
+
+    Examples of Correct Usage:
+    
+    Right :o:
```
Comment (andrei-epure-sonarsource) [inline dotnet-coverage-test-data-import-guide.md]: I find the red confusing for `right` - can this be a green circle?

File: dotnet-coverage-test-data-import-guide.md [hunk 7]
```diff
@@ -1,117 +1,137 @@
...
+    :x: Don't point files that are not under the project root directory, always generate coverage under this home directry
+        `-d:"sonar.cs.vscoveragexml.reportsPaths=..\coverage\*.xml"`
+
+* Do you know how to get debug logs?
+
+    Add `/d:"sonar.verbose=true"` to the...
+    * SonarScanner.MSBuild.exe `begin` command to get more detailed logs (TODO======>should mention the alternative for Azure Devops integration, where the Scanner for MSBuild is hidden)
+    * `SonarCloudPrepare` task in case you are using Azure DevOps
+    
+        * For example: `SonarScanner.MSBuild.exe begin /k:"MyProject" /d:"sonar.verbose=true"`
```
Comment (andrei-epure-sonarsource) [inline dotnet-coverage-test-data-import-guide.md]: this should be between `SonarScanner.MSBuild.exe `begin`` and ``SonarCloudPrepare` task`

File: dotnet-coverage-test-data-import-guide.md [hunk 8]
```diff
@@ -1,117 +1,137 @@
...
+        `-d:"sonar.cs.vscoveragexml.reportsPaths=C:\users\joe\project\coverage\*.xml"`
+        
+    :x: Don't point files that are not under the project root directory, always generate coverage under this home directry
+        `-d:"sonar.cs.vscoveragexml.reportsPaths=..\coverage\*.xml"`
+
+* Do you know how to get debug logs?
+
+    Add `/d:"sonar.verbose=true"` to the...
+    * SonarScanner.MSBuild.exe `begin` command to get more detailed logs (TODO======>should mention the alternative for Azure Devops integration, where the Scanner for MSBuild is hidden)
+    * `SonarCloudPrepare` task in case you are using Azure DevOps
```
Comment (andrei-epure-sonarsource) [inline dotnet-coverage-test-data-import-guide.md]: here you can provide the YAML and/or screenshot of the step

File: dotnet-coverage-test-data-import-guide.md [hunk 9]
```diff
@@ -1,117 +1,137 @@
...
+    
+        * For example: `SonarScanner.MSBuild.exe begin /k:"MyProject" /d:"sonar.verbose=true"`
+    
+            The parsing and import of code coverage happens during the Scanner for MSBuild `end` command (the Azure Devops `SonarCloudAnalyze` task).
```
Comment (andrei-epure-sonarsource) [inline dotnet-coverage-test-data-import-guide.md]:
```suggestion
The parsing and import of code coverage happens during the Scanner for MSBuild `end` command (the Azure Devops `SonarCloudAnalyze` task).
```

This is valid for both points, so should not be indented.

**Review by andrei-epure-sonarsource — COMMENTED** [review]
overall it looks really nice, great work @joe-tingsanchali-sonarsource !

File: dotnet-coverage-test-data-import-guide.md (hunk 6 above)
Comment (andrei-epure-sonarsource) [reply dotnet-coverage-test-data-import-guide.md]: or even better: ✅

File: dotnet-coverage-test-data-import-guide.md [hunk 10]
```diff
@@ -1,117 +1,137 @@
...
+     ```
+     WARN: The Code Coverage report doesn’t contain any coverage data for the included files.
+     ```
+ 
+    * Verify Indexed Paths Compared to Coverage File Paths
+ (see https://community.sonarsource.com/t/code-coverage-is-not-shown-in-the-sonarcloud-io/15218/6)
+* Do the reports contain valid information for the expected set of files? (give examples)
+
+* Did you specify the correct report path property to import the report at the right step?
+When importing code coverage reports, you need to ensure that you specify the report path with your associated code coverage tool's property for the SonarScanner.MSBuild.exe `begin` command, as detailed in [Test Coverage](https://docs.sonarqube.org/latest/analysis/coverage/):
```
Comment (mickael-caro-sonarsource) [inline dotnet-coverage-test-data-import-guide.md]: We also have a automatic detection and write of 
- sonar.cs.vscoveragexml.reportsPaths
- sonar.cs.vstest.reportsPaths

properties inside the Scanner for .NET, if proper trx / coverage files have been found in specific folders.

File: dotnet-coverage-test-data-import-guide.md [hunk 11]
```diff
@@ -1,117 +1,137 @@
...
+* Are the reports in the right format? (need to check which file type matches which code coverage tool)
+
+* Do the reports even exist?
+ If you see the following, then ensure the report generation for your code coverage tool actually produced a file and has content in it.
+ 
+     ```
+     WARN: The Code Coverage report doesn’t contain any coverage data for the included files.
+     ```
+ 
+    * Verify Indexed Paths Compared to Coverage File Paths
```
Comment (mickael-caro-sonarsource) [inline dotnet-coverage-test-data-import-guide.md]: May be worth adding a note here concerning the /p:PathMap option of MSBuild, which is not supported for coverage files being parsed for SQ / SC

File: dotnet-coverage-test-data-import-guide.md (hunk 10 above)
Comment (joe-tingsanchali-sonarsource) [reply dotnet-coverage-test-data-import-guide.md]: @mickael-caro-sonarsource: Which folders are they? I can add that to make it more clear.

Comment (mickael-caro-sonarsource) [reply dotnet-coverage-test-data-import-guide.md]: IIRC, we look at $(Agent.BuildDirectory)\TestResults first, then we fallback in $(Agent.TempDirectory). This is working only on a Azure DevOps environment.

File: dotnet-coverage-test-data-import-guide.md [hunk 12]
```diff
@@ -1,117 +1,161 @@
...
+    ```/d:"sonar.verbose=true``` or
+    ```-d:"sonar.verbose=true"``` 
+  
+    ✅ Report path is relative to the directory where you run SonarScanner for MSBuild 
+    
+    ```-d:"sonar.cs.vscoveragexml.reportsPaths=coveragexml"```
+    
+    **Wrong** ❌
+    
+    ❌  Don't point at directories but files
```
Comment (andrei-epure-sonarsource) [inline dotnet-coverage-test-data-import-guide.md]:
```suggestion
    ❌  Don't point at directories. **How to fix:** point to files.
```

File: dotnet-coverage-test-data-import-guide.md [hunk 13]
```diff
@@ -1,117 +1,161 @@
...
+    
+    ```-d:"sonar.cs.vscoveragexml.reportsPaths=coverage/"```
+        
+    ❌ Don't use absolute paths, favor paths relative to the project root directory
```
Comment (andrei-epure-sonarsource) [inline dotnet-coverage-test-data-import-guide.md]:
```suggestion
    ❌ Don't use absolute paths. **How to fix:** favor paths relative to the project root directory.
```

File: dotnet-coverage-test-data-import-guide.md [hunk 14]
```diff
@@ -1,117 +1,161 @@
...
+    
+    ```-d:"sonar.cs.vscoveragexml.reportsPaths=C:\users\joe\project\coverage\*.xml"```
+        
+    ❌ Don't point files that are not under the project root directory, always generate coverage under this home directory
```
Comment (andrei-epure-sonarsource) [inline dotnet-coverage-test-data-import-guide.md]:
```suggestion
    ❌ Don't point files that are not under the project root directory. **How to fix:** always generate the coverage report under the directory where you run the command from.
```

**Review by andrei-epure-sonarsource — CHANGES_REQUESTED** [review]
I suggest to make the "Wrong" section more uniform and clear.

For example, when my 🧠 sees "Don't use absolute paths, favor paths relative to the project root directory", it firsts wonders whether "Don't" refers to "favor", too.

File: dotnet-coverage-test-data-import-guide.md [hunk 15]
```diff
@@ -1,117 +1,161 @@
...
+  
+    ```-d:"sonar.cs.vscoveragexml.reportsPaths=..\coverage\*.xml"```
+
+* Do you know how to get debug logs?
+
+    * Add `/d:"sonar.verbose=true"` to the...
+        * SonarScanner.MSBuild.exe `begin` command to get more detailed logs (TODO======>should mention the alternative for Azure Devops integration, where the Scanner for MSBuild is hidden)
```
Comment (andrei-epure-sonarsource) [inline dotnet-coverage-test-data-import-guide.md]: it's the `extraProperties`
```
- task: SonarCloudPrepare@1
  inputs:
    SonarCloud: 'sonarcloud'
    organization: 'joetingsanchali'
    scannerMode: 'MSBuild'
    projectKey: 'joetingsanchali_sonar-scanning-someconsoleapp'
    projectName: 'sonar-scanning-someconsoleapp'
    extraProperties: |
      sonar.cs.vstest.reportsPaths=$(Agent.TempDirectory)\**\*.trx
      sonar.cs.vscoveragexml.reportsPaths=$(Agent.TempDirectory)\**\*.coveragexml
      sonar.verbose=true
```

(I took it from [this awesome answer](https://community.sonarsource.com/t/need-help-for-unit-test-code-coverage-for-sonarcloud/32929/5?u=andrei_epure))

Comment (joe-tingsanchali-sonarsource) [reply dotnet-coverage-test-data-import-guide.md]: That Joe guy sure knows something!

File: dotnet-coverage-test-data-import-guide.md (hunk 3 above)
Comment (joe-tingsanchali-sonarsource) [reply dotnet-coverage-test-data-import-guide.md]: I've studied further the art of SonarCloud vs. SonarQube task differences in Azure DevOps and now I see the difference! Updated!

File: dotnet-coverage-test-data-import-guide.md (hunk 10 above)
Comment (joe-tingsanchali-sonarsource) [reply dotnet-coverage-test-data-import-guide.md]: Good call! I see it in the [Code Coverage](https://docs.sonarqube.org/latest/analysis/scan/sonarscanner-for-msbuild/#header-5) section. I already added in the introduction section, but I'll reference it again here as well:  look in $Build.SourcesDirectory path first then fallback to $Agent.TempDirectory.


> In a Azure DevOps / TFS environment, test files are automatically retrieved following this search
>
>  * Search for .trx files in any TestResults folder located under the $Build.SourcesDirectory path
>  * If not found, then a fallback search is made against $Agent.TempDirectory

**Comment by andrei-epure-sonarsource** [comment]
I'd say this should be merged, as it's the state of the community guide.
And when we want to discuss / suggest something else, we can open a PR.

**Review by andrei-epure-sonarsource — APPROVED** [review]

**Pull request merged**